File: 0200-number-of-islands/0200-number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]
        m = len(grid)
        n = len(grid[0])
        res = 0
        
        def dfs(x, y):
            
            grid[x][y] = '0'
            
            
            for d in directions:
                newX = x + d[0]
                newY = y + d[1]
                
                if 0 <= newX < m and 0 <= newY < n and grid[newX][newY] == '1':
                    dfs(newX, newY)
        
        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1':
                    res += 1 
                    dfs(i, j)
        
        return res

File: 0200-number-of-islands/0200-number-of-islands.cs
public class Solution {
    int m = 0;
    int n = 0;
    public int NumIslands(char[][] grid) {
        m = grid.Length;
        n = grid[0].Length;
        int res = 0;
        void Dfs(char[][] grid, int x, int y) {
            grid[x][y] = '0';
            Tuple<int, int>[] directions = {Tuple.Create(-1, 0), Tuple.Create(1, 0), Tuple.Create(0, -1), Tuple.Create(0, 1)};
            foreach (var d in directions) {
               var newX = x + d.Item1;
               var newY = y + d.Item2;
               if (0 <= newX && newX < m && 0 <= newY && newY < n && grid[newX][newY] == '1') {
                   Dfs(grid, newX, newY);
               }
            }
        }
    
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == '1') {
                    res ++;
                    Dfs(grid, i, j);
                }
            }
        }
        return res;
    }
}
    
    